### DISCUSSION TITLE
Discarding typedef for enums?

### DISCUSSION THREAD

**Discussion by Anirban166** [nasa/fprime#1566] [General]
If we are dealing with C++ and not C at any point (assuming, as the deployments have to do with instances of the components or C++ classes), can we discard the `typedef` part when declaring enums? ([e.g.](https://github.com/nasa/fprime/blob/776af784414b8df7c148845dfca8a1956e5ac4cb/Os/Queue.hpp#L27))

(Given that C++ can achieve the goal of discarding the prepended use of `enum` keyword without explicitly creating an alias)

**Comment by astroesteban** [comment]
I wonder how much effort it would be to change those to `enum class` types for added type safety 🤔

**Reply by LeStarch** [reply]
`enum class` is one of those C++11 features that can cause issues on incomplete and old compiler versions.

**Reply by Anirban166** [reply]
This reminds me of `constexpr`, which although some places could find use in fprime, might not be suitable given it's the one from C++11.

**Comment by Anirban166** [comment]
> I wonder how much effort it would be to change those to `enum class` types for added type safety 🤔

That would be better yes, and if we are concerned both about type safety *and* it being strongly-typed as well, then typed enums would be more appropriate:
```cpp
enum enumName: <dataType> {  memberA, memberB, ... };
```
In addition to the assignment of a specific type, this would also avoid the slight verboseness of scoping that comes when using members (`enumName::memberB` for instance) from an enum class.

**Comment by LeStarch** [comment]
We could replace them with enums of the form:

```
enum Something {
   val=1...
}
```
without major change to the code.  These `typedef enum {} name;` constructs come from old C style programming that we are moving away from.

**Reply by Anirban166** [reply]
Would you like me to do the replacements? 
The changes would be fairly straightforward I think, but I wonder if all of the members in the enumeration would need to get assigned a value, like: (considering the [e.g.](https://github.com/nasa/fprime/blob/776af784414b8df7c148845dfca8a1956e5ac4cb/Os/Queue.hpp#L27)) 
```cpp
enum QueueStatus {
                  QUEUE_OK = 1, //!<  message sent/received okay
                  ...,
                  QUEUE_UNKNOWN_ERROR = 10 //!<  Unexpected error; can't match with returns
                 };
```
or if only some of them should have values?

**Reply by LeStarch** [reply]
The rule we use is:

1. All members are explicitly assigned a value **--or--**
2. All members are left to have the compiler assigned a value.

In this case, none of them have values, so we can just write:

```
enum QueueStatus {
            QUEUE_OK, //!<  message sent/received okay
            QUEUE_NO_MORE_MSGS, //!<  If non-blocking, all the messages have been drained.
            QUEUE_UNINITIALIZED, //!<  Queue wasn't initialized successfully
            QUEUE_SIZE_MISMATCH, //!<  attempted to send or receive with buffer too large, too small
            QUEUE_SEND_ERROR, //!<  message send error
            QUEUE_RECEIVE_ERROR, //!<  message receive error
            QUEUE_INVALID_PRIORITY, //!<  invalid priority requested
            QUEUE_EMPTY_BUFFER, //!<  supplied buffer is empty
            QUEUE_FULL, //!< queue was full when attempting to send a message
            QUEUE_UNKNOWN_ERROR //!<  Unexpected error; can't match with returns
};
```

**Reply by LeStarch** [reply]
I am happy to review changes like this, but please try to do one enum or one file per PR so that I can easily perform the review!

**Reply by Anirban166** [reply]
> The rule we use is:
> 
> 1. All members are explicitly assigned a value **--or--**
> 2. All members are left to have the compiler assigned a value.
> 
> In this case, none of them have values, so we can just write:
> 
> ```
> enum QueueStatus {
>             QUEUE_OK, //!<  message sent/received okay
>             QUEUE_NO_MORE_MSGS, //!<  If non-blocking, all the messages have been drained.
>             QUEUE_UNINITIALIZED, //!<  Queue wasn't initialized successfully
>             QUEUE_SIZE_MISMATCH, //!<  attempted to send or receive with buffer too large, too small
>             QUEUE_SEND_ERROR, //!<  message send error
>             QUEUE_RECEIVE_ERROR, //!<  message receive error
>             QUEUE_INVALID_PRIORITY, //!<  invalid priority requested
>             QUEUE_EMPTY_BUFFER, //!<  supplied buffer is empty
>             QUEUE_FULL, //!< queue was full when attempting to send a message
>             QUEUE_UNKNOWN_ERROR //!<  Unexpected error; can't match with returns
> };
> ```

That's reasonable - all or nothing!
And gotcha, we are going with the default case here where the compiler assigns each member with an integer, starting from 0 and incrementing by one.

> I am happy to review changes like this, but please try to do one enum or one file per PR so that I can easily perform the review!

I'll follow that from now on (keeping in mind for all future PRs), for sure!

**Reply by Anirban166** [reply]
@LeStarch Michael, would it be better to do files per fprime subdirectory (wrt the root) for these type of changes? (or just for this specific change?)

Casually thinking about it, but it seems like there would be too many PRs in a repetitive fashion (even for the changes itself) if I were to do one file per PR for this.
For instance, the remaining files under the `OS` directory which would be applicable for this change are [WatchdogTimer.hpp](https://github.com/nasa/fprime/blob/1f7b07da365bc23e10df4184c35d83147b4f1a1e/Os/WatchdogTimer.hpp#L10), [File.hpp](https://github.com/nasa/fprime/blob/1f7b07da365bc23e10df4184c35d83147b4f1a1e/Os/File.hpp#L14), [Directory.hpp](https://github.com/nasa/fprime/blob/1f7b07da365bc23e10df4184c35d83147b4f1a1e/Os/Directory.hpp#L14), [Task.hpp](https://github.com/nasa/fprime/blob/1f7b07da365bc23e10df4184c35d83147b4f1a1e/Os/Task.hpp#L19), [FileSystem.hpp](https://github.com/nasa/fprime/blob/1f7b07da365bc23e10df4184c35d83147b4f1a1e/Os/FileSystem.hpp#L15), [ValidateFile.hpp](https://github.com/nasa/fprime/blob/1f7b07da365bc23e10df4184c35d83147b4f1a1e/Os/ValidateFile.hpp#L27) and [ValidateFileCommon.hpp](https://github.com/nasa/fprime/blob/1f7b07da365bc23e10df4184c35d83147b4f1a1e/Os/ValidateFileCommon.cpp#L117) - now that would be 7 more PRs to just have this minor change done for all files under `OS/`!

**Comment by Anirban166** [comment]
It might also be better to make that small change in the .tmpl files that account for the template structure or the code segment in the autocoder (assuming<sup>*</sup> that to be the source that generates them, could be code in FPP too ig) which produces the 'typedef enum' constructs for the .cpp/.hpp files. (to keep things consistent for future autogenerated C++ code if this change is considered across all files)

The generation for the enum constructs (along with other model-based autocoder-emitted code) in the current version of fprime appears to be from FPP to XML (stored somewhere in the build cache), and then XML to C++, parsed and generated through the Python scripts or the autocoders (although it seems there have been efforts to remove the XML in between, and the FPP-based model itself would be used to directly generate the enums in the C++ files in the long run - like not integrated into the fprime build system yet, but not too far off maybe?)

While going through the Math Component tutorial, I noticed that the .fpp [file specification](https://nasa.github.io/fprime/Tutorials/MathComponent/Tutorial.html#The-MathOp-Type_Construct-the-FPP-Model) included just an 'enum', but for what I saw, the automatically generated header file `MathOpEnumAc.hpp` (after building the model) had a 'typedef enum' for the same enum schema (and from the looks of it, it seems to be following the [template (hpp)](https://github.com/nasa/fprime/blob/7bf79db61b1b63d60a624328dc909ee62869e8ba/Autocoders/Python/src/fprime_ac/generators/templates/enums/enum_hpp.tmpl) for it (and the other .tmpl files with 'typedef enum' constructs that I traced are [cpp](https://github.com/nasa/fprime/blob/7bf79db61b1b63d60a624328dc909ee62869e8ba/Autocoders/Python/src/fprime_ac/generators/templates/component/cpp.tmpl), [enum_hpp](https://github.com/nasa/fprime/blob/7bf79db61b1b63d60a624328dc909ee62869e8ba/Autocoders/Python/src/fprime_ac/generators/templates/component/hpp.tmpl), [array_hpp](https://github.com/nasa/fprime/blob/7bf79db61b1b63d60a624328dc909ee62869e8ba/Autocoders/Python/src/fprime_ac/generators/templates/arrays/array_hpp.tmpl), [SerialH](https://github.com/nasa/fprime/blob/7bf79db61b1b63d60a624328dc909ee62869e8ba/Autocoders/Python/src/fprime_ac/generators/templates/serialize/namespaceSerialH.tmpl) and [namespacePortH](https://github.com/nasa/fprime/blob/7bf79db61b1b63d60a624328dc909ee62869e8ba/Autocoders/Python/src/fprime_ac/generators/templates/port/namespacePortH.tmpl) - just for my reference). <sup>*</sup>I'm not sure what generates that part of the .cpp/.hpp code, but I'm guessing it's the autocoder since the fpp specs didn't have the keyword 'typedef'!

**Reply by LeStarch** [reply]
You are likely right.  We could take a look at the old Python autocoder, but it is slowly being changed over to FPP anyway.  So these constructs should disappear over time.
